### ISSUE TITLE
Expose `KnownSymbol`'s method and `SSymbol`

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/core-libraries-committee#85]
This is a proposal to export `KnownSymbol`'s `symbolSing` method, which is currently internal to `GHC.TypeLits`, along with the associated `SSymbol` data type. This proposal was spun out of [GHC#15183](https://gitlab.haskell.org/ghc/ghc/-/issues/15183), which is part of a larger effort to [clean up some uses of `unsafeCoerce` in the Haskell ecosystem](https://gitlab.haskell.org/ghc/ghc/-/issues/21568) involving classes like `KnownSymbol`.

# Motivation

## The `KnownSymbol` class

Currently, `GHC.TypeLits` provides a [`KnownSymbol`](https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-TypeLits.html#t:KnownSymbol) class:

```hs
module GHC.TypeLits
  ( KnownSymbol
  , ...
  ) where

class KnownSymbol (n :: Symbol) where
  symbolSing :: SSymbol n

newtype SSymbol (s :: Symbol) = UnsafeSSymbol String

...
```

This class is treated specially by GHC, as GHC automatically solves `KnownSymbol` constraints during typechecking. Only the class is exported from `GHC.TypeLits`; its `symbolSing` method, as well as the associated `SSymbol` data type, are kept internal to the module.

The idea behind `SSymbol s` is that the `Symbol` `s` is the type-level equivalent of the underlying `String`. For instance, so one could have `UnsafeSSymbol "a" :: SSymbol "a"` or `UnsafeSSymbol "b" :: SSymbol "b"`. The internals of `GHC.TypeLits` are careful to ensure that `UnsafeSSymbol` is never used to construct an `SSymbol` whose `Symbol` type disagrees with the underlying `String`.

## The `reflection` library

For most use cases, keeping `SSymbol` internal to the module is not an issue, as `GHC.TypeLits` defines an API that allows interfacing with `KnownSymbol` by other means. There is one use case for which the current API is not sufficient, however: the [`reflection`](https://hackage.haskell.org/package/reflection) library. In particular, `reflection` offers a [`reifySymbol`](https://hackage.haskell.org/package/reflection-2.1.6/docs/Data-Reflection.html#v:reifySymbol) function:

```hs
reifySymbol :: forall r. String -> (forall (n :: Symbol). KnownSymbol n => Proxy n -> r) -> r
```

This function allows one to construct a `KnownSymbol` constraint from a `String`. It is possible to define `reifySymbol` to define the API provided in `GHC.TypeLits`:

```hs
reifySymbol :: forall r. String -> (forall (s :: Symbol). KnownSymbol s => Proxy s -> r) -> r
reifySymbol s f =
  case someSymbolVal s of
    SomeSymbol p -> f p
```

However, this implementation is not as performant as it could be, as it requires pattern-matching on an intermediate `SomeSymbol` data structure. `reflection` instead defines `reifySymbol` using `unsafeCoerce`:

```hs
reifySymbol :: forall r. String -> (forall (s :: Symbol). KnownSymbol s => Proxy s -> r) -> r
reifySymbol s k = unsafeCoerce (MagicSymbol k :: MagicSymbol r) s Proxy

newtype MagicSymbol r = MagicSymbol (forall (s :: Symbol). KnownSymbol s => Proxy s -> r)
```

This use of `unsafeCoerce` only works because when GHC compiles a class containing a single method (e.g., `KnownSymbol`), the class is treated as if it were a newtype, so the `unsafeCoerce` is effectively coercing underneath a newtype. For more technical details on how this works, refer to [one](https://www.schoolofhaskell.com/user/thoughtpolice/using-reflection) of these [two](https://www.tweag.io/posts/2017-12-21-reflection-tutorial.html) tutorials on the `reflection` library. Knowing these technical details are not required to understand the rest of this proposal.

## An alternative to `unsafeCoerce`

While `reifySymbol`'s current, `unsafeCoerce`-based implementation does work, it may not be future-proof. In particular, it crucially relies on the assumption that single-method classes are treated like newtypes when compiled. This assumption may not hold in the future depending on the outcome of [this discussion](https://gitlab.haskell.org/ghc/ghc/-/issues/20897), which proposes to change the representation of single-method classes. If the representation of single-method classes switches away from that of newtypes, then the `unsafeCoerce` trick would likely no longer work.

In anticipation of such a possible future, `GHC.Exts` now offers a [`withDict`](https://hackage.haskell.org/package/ghc-prim-0.9.0/docs/GHC-Magic-Dict.html#v:withDict) primitive, which was designed with use cases like `reflection`'s in mind. `withDict` provides a way to build a constraint for a single-method class, but unlike `unsafeCoerce`, `withDict` will throw a compile-time error if improperly used. For example, one can define a function that converts from an `SSymbol` to a `KnownSymbol` constraint like so:

```hs
withKnownSymbol :: forall s r. SSymbol s -> (KnownSymbol s => r) -> r
withKnownSymbol s k = withDict @(KnownSymbol s) @(SSymbol s) s k
```

Using this, one can define `reifySymbol` like so:

```hs
reifySymbol :: forall r. String -> (forall (s :: Symbol). KnownSymbol s => Proxy s -> r) -> r
reifySymbol str k = withKnownSymbol (UnsafeSSymbol str) (k Proxy)
```

Being able to use `withDict` to define the `reflection` library like this is one of the goals of [this GHC issue](https://gitlab.haskell.org/ghc/ghc/-/issues/21568).

## The API isn't expressive enough

Unfortunately, it is not possible to write the `withDict`-based version of `reifySymbol` using the API that `GHC.TypeLits` currently provides. As noted above, `GHC.TypeLits` does not export the `SSymbol` data type, so it is neither possible to define `withKnownSymbol` (which uses `SSymbol`) nor `reifySymbol` (with uses the `UnsafeSSymbol` data constructor). Indeed, we certainly do not want to export `UnsafeSSymbol`, as it is possible to construct completely invalid values with it. (Think `UnsafeSSymbol "a" :: SSymbol "b"`.) On the other hand, not exporting `SSymbol` at all makes it impossible to define combinators such as `withKnownSymbol`, which we argue is too restrictive.

## A brief aside about `KnownNat` and `KnownChar`

While the running example throughout this proposal uses `KnownSymbol`, similar issues apply to the `KnownNat` and `KnownChar` constraints, which are to `Nat` and `Char` what `KnownSymbol` is to `Symbol`. Just as `KnownSymbol` has a hidden method involving the `SSymbol` type, `KnownNat` has a hidden method involving the `SNat` type, and `KnownChar` has a hidden method involving the `SChar` type.

# Proposal

We propose to extend the API offered by `GHC.TypeLits` in the following ways:

1. Export the `symbolSing` method of `KnownSymbol`.
2. Export the `SSymbol` data type, but _not_ its data constructor `UnsafeSSymbol`.
3. Define the following `Show`, `TestEquality`, and `TestCoercion` instances for `SSymbol`:

   ```hs
   instance Show (SSymbol n) where
     showsPrec p (UnsafeSSymbol n)
       = showParen (p > appPrec)
         ( showString "SSymbol @"
           . showsPrec appPrec1 n
         )

   instance TestEquality SSymbol where
     testEquality (UnsafeSSymbol x) (UnsafeSSymbol y)
       | x == y    = Just (unsafeCoerce Refl)
       | otherwise = Nothing

   instance TestCoercion SSymbol where
     testCoercion x y = fmap (\Refl -> Coercion) (testEquality x y)
   ```

   These instances are directly inspired by the [`singletons`](https://hackage.haskell.org/package/singletons) library, which provide similar instances for singleton types such as `SSymbol`.
3. Export the following function, which extracts the underlying `String` from an `SSymbol`:

   ```hs
   fromSSymbol :: SSymbol s -> String
   fromSSymbol (UnsafeSSymbol s) = s
   ```
4. Export the following function, which allows constructing an `SSymbol` from a `String`:

   ```hs
   withSomeSSymbol :: forall rep (r :: TYPE rep). String -> (forall s. SSymbol s -> r) -> r
   withSomeSSymbol s k = k (UnsafeSSymbol s)
   ```
5. Export the following function, which builds a `KnownSymbol` constraint from an `SSymbol`:

   ```hs
   withKnownSymbol :: forall s rep (r :: TYPE rep). SSymbol s -> (KnownSymbol s => r) -> r
   withKnownSymbol s k = withDict @(KnownSymbol s) @(SSymbol s) s k
   ```

   This is defined to be representation polymorphic for symmetry with the existing [`withTypeable`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Type-Reflection.html#v:withTypeable) function, which serves an analogous role in `Data.Typeable`.
6. Export the following `SSymbol` pattern synonym, which allows matching on an `SSymbol` to bring a `KnownSymbol` constraint into scope:

   ```hs
   pattern SSymbol :: forall s. () => KnownSymbol s => SSymbol s
   pattern SSymbol <- (knownSymbolInstance -> KnownSymbolInstance)
     where SSymbol = symbolSing

   -- Internal to GHC.TypeLits
   data KnownSymbolInstance (s :: Symbol) where
     KnownSymbolInstance :: KnownSymbol n => KnownSymbolInstance n

   -- Internal to GHC.TypeLits
   knownSymbolInstance :: SSymbol s -> KnownSymbolInstance s
   knownSymbolInstance ss = withKnownSymbol ss KnownSymbolInstance
   ```

   This is for symmetry with the existing [`TypeRep`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Type-Reflection.html#v:TypeRep) pattern synonym, which serves an analogous role in `Data.Typeable`.

We also propose similar changes involving `KnownNat` and `KnownChar` so that `SNat` and `SChar` are exposed and are given corresponding APIs. These changes would be nearly identical to the ones shown above, except with `Symbol` replaced with `Nat`/`Char`, `String` replaced with `Natural`/`Char`, etc.

The primary benefit of these API additions is that `reifySymbol` from the `reflection` library can now be defined like this, without `unsafeCoerce`:

```hs
reifySymbol :: forall r. String -> (forall (s :: Symbol). KnownSymbol s => Proxy s -> r) -> r
reifySymbol s k = withSomeSSymbol s $ \(ss :: SSymbol s) -> withKnownSymbol ss (k @s Proxy)
```

A secondary benefit of this proposal is that the [`singletons`](https://hackage.haskell.org/package/singletons) library would benefit from having access to `SSymbol`. Currently, `singletons` defines [its own version of `SSymbol`](https://hackage.haskell.org/package/singletons-base-3.1.1/docs/GHC-TypeLits-Singletons.html#t:SSymbol), which uses an existential `KnownSymbol` constraint. If `singletons` could build on top of the `SSymbol` from `GHC.TypeLits`, it would shave off one layer of indirection by avoiding the existential constraint.

# Alternatives

We have proposed several API additions to `GHC.TypeLits`, but and there is room for bikeshedding on what the exact design should look like. Aside from the usual bikeshedding topics, such as what names to give top-level functions, some possible tweaks one could make include:

1. Instead of exporting `SNat` data type, we could continue to keep it internal to `GHC.TypeLits` and instead offer a function with the same type signature as `reifySymbol`. That way, the `reflection` library would a way to define `reifySymbol`-sans-`unsafeCoerce` without needing to expose `SSymbol` as part of the public API of `GHC.TypeLits`.

   While this is doable, we think it is more reasonable to just export `SSymbol`. After all, [users are asking for it](https://gitlab.haskell.org/ghc/ghc/-/issues/15183), and exposing the additional functionality will not impact existing code using `GHC.TypeLits`. Moreover, libraries such as `singletons` would be able to build on top of `singletons`, which is another reason in favor of exporting it.
2. Instead of exporting the `symbolSing` method of `KnownSymbol`, we could keep `symbolSing` hidden and instead provide a top-level function with the same type signature as `symbolSing`. While possible, we argue that this level of caution is not needed. Even if `symbolSing` is exposed, one is not allowed to define custom instances of `KnownSymbol` anyway due to its special treatment in the compiler, so exporting `symbolSing` does not pose any additional risk.
3. Strictly speaking, we do not need to provide the `withKnownSymbol` function, as users could define it themselves using `withDict`. That being said, it seems reasonable to provide it as a convenience.
4. The `Show` instance proposed above adopts the pretty-printing conventions used in the `singletons` library. Alternatively, we could derive a `Show` instance for `SSymbol`, but derived `Show` instances necessarily leak out some information about the structure of the data type. For this reason, we would prefer to define the `Show` instance by hand.

**Comment by mixphix** [comment]
Thanks for this detailed proposal, @RyanGlScott! The `withDict` function and its compiler-related magics are obscure to many of us who are not regular GHC contributors or heavy users of type-level machinery. I appreciate the many references you gave to help bring us up to speed on the minutiae.

During my reading I noticed issue [#21575](https://gitlab.haskell.org/ghc/ghc/-/issues/21575), which you reference in [the ticket you linked](https://gitlab.haskell.org/ghc/ghc/-/issues/21568) at the end of the section about `unsafeCoerce`. You reference it as part 5 of the plan for `withDict`, and to my eye it seems unresolved: it interacts poorly with GHC's specialiser and allows the inappropriate swapping of class dictionaries. How certain are you that using it in this case will avoid compilation errors?

Due to the existing symmetries you mention with `Data.Typeable`, I see no other reason to oppose this proposal. Besides, I'm in favour of facilitating dependently-typed programming.

**Comment by RyanGlScott** [comment]
> During my reading I noticed issue [#21575](https://gitlab.haskell.org/ghc/ghc/-/issues/21575), which you reference in [the ticket you linked](https://gitlab.haskell.org/ghc/ghc/-/issues/21568) at the end of the section about `unsafeCoerce`. You reference it as part 5 of the plan for `withDict`, and to my eye it seems unresolved: it interacts poorly with GHC's specialiser and allows the inappropriate swapping of class dictionaries. How certain are you that using it in this case will avoid compilation errors?

A very good question. [The issue that you mention](https://gitlab.haskell.org/ghc/ghc/-/issues/21568) only arises under very particular circumstances:

1. There is a dictionary `c1 :: C`,
2. There is another dictionary `c2 :: C` such that the value of `c2` is different from the value of `c1`, and
3. GHC's specialiser swaps out `c2` for `c1`, resulting in a change of program behavior.

Depending on the choice of class `C`, however, it is possible that (2) can never arise. For instance, we already use `withDict` in the `base` library to implement [`withTypeable`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Type-Reflection.html#v:withTypeable). In the case of `withTypeable`, `C` is `Typeable`, where `Typeable` is defined as:

```hs
class Typeable a where
  typeRep# :: TypeRep a
```

Why doesn't (2) ever arise here? It's because `TypeRep` is a _singleton_ type. That is, given a fixed type `a`, there is only ever a single unique value¹ that can inhabit type `TypeRep a`. To put it another way, if we have `c1, c2 :: TypeRep a`, then it is always safe for GHC's specialiser to swap out `c2` for `c1`, as `c1` and `c2` are guaranteed to have the same value.

I claim that `SSymbol` can be made into a singleton type in the same vein as `TypeRep`, making `SSymbol` safe for use with `withDict`. Certainly, `SSymbol` is not a singleton type in general: if one had unrestricted access to the `UnsafeSSymbol` data constructor, then one could have `UnsafeSSymbol "a"` and `UnsafeSSymbol "b"` both be of type `SSymbol "a"`, which would break the singleton type invariant. Therefore, it is important that programmers are restricted in the ways that they can create `SSymbol` values such that `SSymbol` is guaranteed to uphold the invariant.

The API that I propose is designed to uphold this invariant. The `UnsafeSSymbol` data constructor is not exported, and the only way to manufacture a fresh `SSymbol s` value is by way of the `withSomeSSymbol` function. The type of `withSomeSSymbol` quantifies the `s` in a higher-rank type, so each `SSymbol` value that `withSomeSSymbol` creates will have a distinct type is not equal to any other `SSymbol` type. As a result, if we have `withSomeSSymbol s (\sym -> ...)`, then GHC's specialiser cannot swap out `sym` for a different `SSymbol` value, as the types will not be the same.²

In summary, I'm quite certain that `withDict` won't pose any issues when it comes to `SSymbol` and friends. Let me know if you think the any parts of the proposal should be updated to clarify things.

¹ Besides `⊥`, but [it is fine to reason modulo `⊥`](https://www.cs.ox.ac.uk/jeremy.gibbons/publications/fast+loose.pdf).
² More precisely, this works provided that `withSSomeSymbol` is marked as `NOINLINE`. See [this Note](https://gitlab.haskell.org/ghc/ghc/-/blob/161a6f1fd62e797e978e7808a5f567fefa123f16/libraries/base/GHC/TypeNats.hs#L95-150) for the technical reasons behind this.

**Comment by RyanGlScott** [comment]
While writing a proof-of-concept implementation of this patch, I discovered one minor thing that the proposal above does not address: there are actually _two_ different module APIs for `KnownNat`. There is `GHC.TypeLits`, an older module where `KnownNat` operations involve `Integer`s, e.g.,

```hs
natVal :: KnownNat n => proxy n -> Integer
someNatVal :: Integer -> Maybe SomeNat
```

There is also `GHC.TypeNats`, a newer module where `KnownNat` operations involve `Natural` numbers rather than `Integer`s, e.g.,

```hs
natVal :: KnownNat n => proxy n -> Natural
someNatVal :: Natural -> SomeNat
```

The API in `GHC.TypeNats` is what most new code uses because type-level numbers are always non-negative, and thus negative `Integer` values do not have type-level counterparts. This is why `GHC.TypeLits.someNatVal` has a `Maybe` in its return type, as it returns `Nothing` if its argument is negative.

When writing this proposal, I had `GHC.TypeNats` in mind. There are two functions from this proposal that would need to mention `Natural`s when added to `GHC.TypeNats`:

```hs
fromSNat :: SNat n -> Natural
withSomeSNat :: Natural -> (forall n. SNat n -> r) -> r
```

However, to keep the symmetry between the APIs in `GHC.TypeLits` and `GHC.TypeNats`, we should also add functions to `GHC.TypeLits` as well. While most of the new `GHC.TypeNats` functions can simply be re-exported from `GHC.TypeLits`, the `fromSNat` and `withSomeSNat` functions would need to be altered to mention `Integer` instead of `Natural`. I propose these type signatures:

```hs
-- Invariant: the returned Integer is always non-negative.
fromSNat :: SNat n -> Integer

-- If the Integer argument is negative, return the first `r` argument.
-- Otherwise, invoke the continuation.
withSomeSNat :: Integer -> r -> (forall n. SNat n -> r) -> r
```

There is a separate question of whether we should continue to maintain these parallel APIs in `GHC.TypeLits` and `GHC.TypeNats`, but attempting to unify the two would require a breaking API change. As such, I will defer this question to a separate proposal.

**Comment by goldfirere** [comment]
Just to provide some context: The current design arose as a compromise. In the early days of singletons, Iavor and I couldn't quite settle on the right design. He was working on the type-lits stuff, while I was working on the singletons library. In order to make progress, it was decided that the built-in type-lits interface would be as small as possible to give access to GHC-internal primitives, while the singletons library would have enough facility to make the type-lits stuff more ergonomic. In some sense, we didn't want to commit to any particular design before gaining experience with real users. I think plenty of time has passed now, and I'm happy to let Ryan's expertise and experience guide us here.

**Comment by RyanGlScott** [comment]
An administrative question: is there a formal process by which I should submit this proposal for consideration by the CLC? I ask since there have been no new comments on the proposal for a week, so I think this has reached a state of quiescence.

I also have a GHC branch [here](https://gitlab.haskell.org/RyanGlScott/ghc/-/commit/490d04f495b95205f3574c8323db34bfe1d3542f) if you'd like to see what the full implementation of this proposal would look like.

**Comment by mixphix** [comment]
No, this comment is enough indication that you'd like us to cast a vote. Typically we ask that breaking changes submit an impact assessment; this usually involves building stackage with the changes applied to `base` and seeing how many packages break. In your case I don't imagine the number will be very high.

**Comment by RyanGlScott** [comment]
OK. While this proposal doesn't have any breaking changes per se (it only exposes new functionality), it is possible that some of the names that I have chosen would clash would existing library code. I could imagine `singletons`-related libraries being affected, for instance, as they define their own `SNat` type. I agree that an impact analysis would be wise.

Do you (for any definition of "you") have any tooling for performing these sorts of impact analyses? I was imagining that I would apply the patch [here](https://gitlab.haskell.org/RyanGlScott/ghc/-/commit/490d04f495b95205f3574c8323db34bfe1d3542f) to a stable GHC version's source code, build it, and then construct a `.cabal` file containing everything in a recent Stackage LTS. This process might be a bit fiddly, so I'm wondering if others have tried to script it to any degree.

**Comment by Bodigrim** [comment]
@RyanGlScott please check https://github.com/Bodigrim/clc-stackage

**Comment by RyanGlScott** [comment]
Thanks, @Bodigrim. I have some questions about how to use it:

1. This appears to be driven by a Nix configuration. Where in the configuration do I specify the patched version of GHC that I want to use for testing?
2. After that, what command(s) do I run to make it build everything in Stackage?

**Comment by Bodigrim** [comment]
You can ignore all Nix parts, just `cabal build -w ghc-HEAD` it.

**Comment by ekmett** [comment]
I only regret that I have but one :+1: to give. The current state leads to so many oh-so-many `unsafeCoerce`s in my code.

**Comment by RyanGlScott** [comment]
Upon further thought, I've realized that testing against Stackage is somewhat tricky. This is because the changes in this proposal require `withDict`, which needs GHC 9.4 or later. On the other hand, Stackage Nightly only supports GHC 9.2 at present. One could imagine manually pruning the subset of Stackage Nightly that currently supports GHC 9.4, but that sounds like a pretty tedious experience.

Instead of Stackage Nightly, would it be acceptable to target [`head.hackage`](https://gitlab.haskell.org/ghc/head.hackage) for an impact assessment? `head.hackage` curates a list of package constraints [here](https://ghc.gitlab.haskell.org/head.hackage/cabal.constraints), along with some additional packages [here](https://gitlab.haskell.org/ghc/head.hackage/-/blob/b078702bdbb8f54ff96937b4ce165860c2e1368e/ci/config.sh#L95-116), and they are all known to build against GHC 9.4. It certainly doesn't provide as many packages as Stackage does, but it's a pretty large number of packages nonetheless. It also includes `singletons`, which is something that I definitely want to test.

**Comment by Bodigrim** [comment]
@RyanGlScott I think you can find all potentially affected packages via https://hackage-search.serokell.io/?q=%5CbsymbolSing%5Cb and https://hackage-search.serokell.io/?q=%5CbSSymbol%5Cb

**Comment by RyanGlScott** [comment]
Ah, if you're OK with me only building the packages that mention the new API additions by name, then I'll just do that. That makes my job a lot easier.

**Comment by Bodigrim** [comment]
This is not a breaking change in PVP terms, am I correct? We just export couple of new entities from `GHC.*` module. The process does not insist on impact assessment in such cases at all, so a cursory glance would be enough.

**Comment by RyanGlScott** [comment]
> This is not a breaking change in PVP terms, am I correct? We just export couple of new entities from GHC.* module.

Correct.

> The process does not insist on impact assessment in such cases at all, so a cursory glance would be enough.

Even better! I'll do just that.

**Comment by RyanGlScott** [comment]
Of the 117 potentially affected packages that https://hackage-search.serokell.io turned up, I found five packages that (1) build successfully with GHC 9.4 and (2) do not build a patched version of GHC 9.4 (using the patch from [here](https://gitlab.haskell.org/RyanGlScott/ghc/-/commit/490d04f495b95205f3574c8323db34bfe1d3542f)). As one might expect, all of the build failures are due to these libraries defining their own functions with the same names as the API additions proposed above.

## `parameterized-utils`

This package defines its own [`withKnownNat`](https://hackage.haskell.org/package/parameterized-utils-2.1.5.0/docs/Data-Parameterized-NatRepr.html#v:withKnownNat) function. Here is a patch to avoid the issue:

<details>

```diff
diff --git a/src/Data/Parameterized/NatRepr.hs b/src/Data/Parameterized/NatRepr.hs
index 64465f9..e71d0a2 100644
--- a/src/Data/Parameterized/NatRepr.hs
+++ b/src/Data/Parameterized/NatRepr.hs
@@ -134,7 +134,9 @@ import Data.Data
 import Data.Type.Equality as Equality
 import Data.Void as Void
 import Numeric.Natural
-import GHC.TypeNats as TypeNats
+import GHC.TypeNats ( KnownNat, Nat, SomeNat(..)
+                    , type (+), type (-), type (*), type (<=)
+                    , someNatVal )
 import Unsafe.Coerce

 import Data.Parameterized.Axiom
```
</details>

## `sbv`

This package defines its own [`SChar`](https://hackage.haskell.org/package/sbv-9.0/docs/Data-SBV.html#t:SChar) type. Here is a patch to avoid the issue:

<details>

```diff
diff --git a/Data/SBV.hs b/Data/SBV.hs
index 907913d..69cdae9 100644
--- a/Data/SBV.hs
+++ b/Data/SBV.hs
@@ -464,7 +464,7 @@ import Data.SBV.Control.Types (SMTReasonUnknown(..), Logic(..))
 import qualified Data.SBV.Utils.CrackNum as CN

 import Data.Proxy (Proxy(..))
-import GHC.TypeLits
+import GHC.TypeLits (KnownNat, type (+), type (-), type (<=))

 import Prelude hiding((+), (-), (*)) -- to avoid the haddock ambiguity

@@ -939,7 +939,7 @@ the 'Data.SBV.Core.Model.Uninterpreted' class.
 -}

 {- $enumerations
-If the uninterpreted sort definition takes the form of an enumeration (i.e., a simple data type with all nullary constructors), then
+If the uninterpreted sort definition takes the form of an enumeration (i.e., a simple data type with all nullary constructors), then
 tou can use the 'mkSymbolicEnumeration' functio to turn it into an enumeration in SMTLib.
 A simple example is:

diff --git a/Data/SBV/Client/BaseIO.hs b/Data/SBV/Client/BaseIO.hs
index 512cee7..6a21f2a 100644
--- a/Data/SBV/Client/BaseIO.hs
+++ b/Data/SBV/Client/BaseIO.hs
@@ -39,7 +39,7 @@ import Data.SBV.Provers.Prover (Provable, SExecutable, ThmResult)
 import Data.SBV.SMT.SMT        (AllSatResult, SafeResult, SatResult,
                                 OptimizeResult)

-import GHC.TypeLits
+import GHC.TypeLits            (ErrorMessage(..), KnownNat, TypeError)
 import Data.Kind

 import Data.Int
diff --git a/Data/SBV/Core/Data.hs b/Data/SBV/Core/Data.hs
index c600e6a..cd5a64a 100644
--- a/Data/SBV/Core/Data.hs
+++ b/Data/SBV/Core/Data.hs
@@ -62,7 +62,7 @@ module Data.SBV.Core.Data
  , QueryState(..), QueryT(..), SMTProblem(..)
  ) where

-import GHC.TypeLits
+import GHC.TypeLits (Nat)

 import GHC.Generics (Generic)
 import GHC.Exts     (IsList(..))
```
</details>

## `singletons-base`

This package defines much of the same API as what is proposed above, including [`SChar`](https://hackage.haskell.org/package/singletons-base-3.1.1/docs/GHC-TypeLits-Singletons.html#t:SChar), [`SNat`](https://hackage.haskell.org/package/singletons-base-3.1.1/docs/GHC-TypeLits-Singletons.html#t:SNat), and [`SSymbol`](https://hackage.haskell.org/package/singletons-base-3.1.1/docs/GHC-TypeLits-Singletons.html#t:SSymbol). Here is a patch to avoid the issue:

<details>

```diff
diff --git a/src/GHC/TypeLits/Singletons/Internal.hs b/src/GHC/TypeLits/Singletons/Internal.hs
index 089ca1b..aa8ef29 100644
--- a/src/GHC/TypeLits/Singletons/Internal.hs
+++ b/src/GHC/TypeLits/Singletons/Internal.hs
@@ -49,7 +49,11 @@ import Data.Singletons.Decide
 import Data.Singletons.TH
 import GHC.Show (appPrec, appPrec1)
 import GHC.Stack (HasCallStack)
-import GHC.TypeLits as TL
+import qualified GHC.TypeLits as TL
+import GHC.TypeLits ( KnownChar, KnownNat, KnownSymbol, Natural
+                    , SomeChar(..), SomeNat(..), SomeSymbol(..), Symbol
+                    , type (^), type (<=?), charVal, sameChar, sameNat
+                    , sameSymbol, someCharVal, someSymbolVal, symbolVal )
 import qualified GHC.TypeNats as TN
 import Unsafe.Coerce

```
</details>

A future version of `singletons-base` will likely want to build on top of the `SSymbol` and friends rather than redefining them, however.

## `typelits-printf`

This package defines its own [`SChar`](https://hackage.haskell.org/package/typelits-printf-0.2.0.0/docs/GHC-TypeLits-Printf.html#t:SChar) type. Here is a patch to avoid the issue:

<details>

```diff
diff --git a/src/GHC/TypeLits/Printf/Internal.hs b/src/GHC/TypeLits/Printf/Internal.hs
index 6cd55f9..88449c2 100644
--- a/src/GHC/TypeLits/Printf/Internal.hs
+++ b/src/GHC/TypeLits/Printf/Internal.hs
@@ -62,7 +62,8 @@ import           Data.Proxy
 import           Data.Symbol.Utils
 import           Data.Word
 import           GHC.OverloadedLabels
-import           GHC.TypeLits
+import           GHC.TypeLits ( ErrorMessage(..), KnownSymbol, Symbol
+                              , TypeError, symbolVal )
 import           GHC.TypeLits.Printf.Parse
 import           Numeric.Natural
 import qualified Data.Text                 as T
diff --git a/src/GHC/TypeLits/Printf/Internal/Parser.hs b/src/GHC/TypeLits/Printf/Internal/Parser.hs
index cf8c6df..fe51a1f 100644
--- a/src/GHC/TypeLits/Printf/Internal/Parser.hs
+++ b/src/GHC/TypeLits/Printf/Internal/Parser.hs
@@ -11,7 +11,8 @@ module GHC.TypeLits.Printf.Internal.Parser where
 import           Data.Kind
 import           Data.Type.Bool
 import           Data.Type.Equality
-import           GHC.TypeLits
+import           GHC.TypeLits ( AppendSymbol, ErrorMessage(..), Nat, Symbol
+                              , TypeError, type (+), type (*) )

 -- | A type synonym for a single-character symbol.  Ideally this would just
 -- be 'Char', but we don't have chars at the type level.  So, if you see
diff --git a/src/GHC/TypeLits/Printf/Parse.hs b/src/GHC/TypeLits/Printf/Parse.hs
index 263ffcb..513e523 100644
--- a/src/GHC/TypeLits/Printf/Parse.hs
+++ b/src/GHC/TypeLits/Printf/Parse.hs
@@ -30,7 +30,7 @@ module GHC.TypeLits.Printf.Parse (

 import           Data.Proxy
 import           Data.Text                           (Text)
-import           GHC.TypeLits hiding                 (natVal)
+import           GHC.TypeLits                        (AppendSymbol, KnownSymbol, Symbol, symbolVal)
 import           GHC.TypeLits.Printf.Internal.Parser
 import           GHC.TypeNats
 import           Numeric.Natural
```
</details>

## `winery`

This package defines an [`SChar`](https://github.com/fumieval/winery/blob/5f073161937e3e58e26558222d367ec185659c22/src/Codec/Winery/Base.hs#L97) data constructor that clashes with the proposed `SChar` pattern synonym. Here is a patch to avoid the issue:

<details>

```diff
diff --git a/src/Codec/Winery/Class.hs b/src/Codec/Winery/Class.hs
index d9a5f65..da96f1b 100644
--- a/src/Codec/Winery/Class.hs
+++ b/src/Codec/Winery/Class.hs
@@ -116,7 +116,7 @@ import Unsafe.Coerce
 import GHC.Float (castWord32ToFloat, castWord64ToDouble)
 import GHC.Natural
 import GHC.Generics
-import GHC.TypeLits
+import GHC.TypeLits (KnownSymbol, symbolVal)

 -- | Serialisable datatype
 --
@@ -1104,4 +1104,4 @@ buildRecordExtractor = buildExtractor . btraverse (fmap Identity)
 -- | Collect extractors for record fields
 bextractors :: forall b. (ConstraintsB b, AllB Serialise b, FieldNamesB b) => b Subextractor
 bextractors = bmapC @Serialise (extractField . getConst) bfieldNames
-{-# INLINABLE bextractors #-}
\ No newline at end of file
+{-# INLINABLE bextractors #-}
```
</details>

**Comment by Bodigrim** [comment]
Thanks, @RyanGlScott, looks good. Basically all build failures are name clashes. 

Since the proposal and the [patch](https://gitlab.haskell.org/RyanGlScott/ghc/-/commit/490d04f495b95205f3574c8323db34bfe1d3542f) are quite extensive, I'll keep the floor open for another week, so that everyone has a chance to read and ask questions. If there are no unresolved questions by the end of the next weekend, I'll trigger a vote.

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal as detailed above. @mixphix @tomjaguarpaw @cgibbard @chessai @emilypi 

----

+1 from me.

**Comment by mixphix** [comment]
+1

**Comment by chessai** [comment]
+1

**Comment by tomjaguarpaw** [comment]
+1

----

Great proposal write up, thanks.

**Comment by emilypi** [comment]
+1

**Comment by cgibbard** [comment]
+1

**Comment by Bodigrim** [comment]
Thanks all, approved unanimously.

**Comment by Bodigrim** [comment]
@RyanGlScott please ping me to approve when you raise an MR.

**Comment by RyanGlScott** [comment]
@Bodigrim, see [GHC!9064](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9064).

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @RyanGlScott   |
| **Status**             |  merged  |
| **`base` version**     | `4.18.0.0`   |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9064 |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | present   |
| **Migration guide**    |  not needed |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂
